File: p312/p312.cpp
#include <iostream>
#include <cstdio>
#include <vector>
using namespace std;

class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
        	return 0;
        if (n == 1)
        	return nums[0];

        nums.insert(nums.begin(),1);
        nums.push_back(1);

        vector<vector<int> > f(n+2,vector<int>(n+2,0));

        for (int i = 1; i <= n; ++i)
        	f[i][i] = nums[i-1]*nums[i]*nums[i+1];
        
        for (int i = 2; i <= n; ++i)
        {
        	for (int j = 1; j <= n-i+1; ++j)
        	{
        		int en = j+i-1;
        		f[j][en] = max(f[j+1][en] + nums[j]*nums[j-1]*nums[en+1],
        						f[j][en-1] + nums[en]*nums[j-1]*nums[en+1]);
        		for (int k = j+1; k < en; ++k)
        			f[j][en] = max(f[j][en],
        				f[j][k-1] + f[k+1][en] + nums[j-1]*nums[k]*nums[en+1]);

        		// printf("%d %d %d\n", j,en,f[j][en]);
        	}
        }

        return f[1][n];
    }
};

int main(int argc, char const *argv[])
{
	Solution s;
	int d[] = {1,2,3};
	vector<int> v(d,d + sizeof(d)/sizeof(int));
	cout << s.maxCoins(v) << endl;
	return 0;
}

File: p312/p312.go
package p312

const maxn = 510

func maxCoins(nums []int) int {
	n := len(nums)

	nums = append([]int{1}, nums...)
	nums = append(nums, 1)

	n += 2
	f := make([][]int, n)
	for i := 0; i < n; i++ {
		f[i] = make([]int, n)
	}

	for l := 3; l <= n; l++ {
		for en := l - 1; en < n; en++ {
			st := en - l + 1

			for k := st + 1; k < en; k++ {
				cur := f[st][k] + f[k][en] + nums[st]*nums[k]*nums[en]
				if cur > f[st][en] {
					f[st][en] = cur
				}
			}
		}
	}

	return f[0][n-1]
}
